pubsub/pubsub_goredis.go

```diff
@@ -83,88 +85,58 @@ func (ps *GoRedisPubSub) Stop() error {
-// started to listen for messages on the topic; each message is sent to all
-// subscribers to the topic.
-func (ps *GoRedisPubSub) NewTopic(ctx context.Context, topic string) Topic {
-	t := &GoRedisTopic{
-		client:      ps.client,
-		topic:       topic,
-		redisSub:    ps.client.Subscribe(ctx, topic),
-		subscribers: make([]chan string, 0),
-		done:        make(chan struct{}),
+func (ps *GoRedisPubSub) Close() {
+	ps.mut.Lock()
+	defer ps.mut.Unlock()
+	for _, sub := range ps.subs {
+		sub.Close()
 	}
+	ps.subs = nil
+	ps.client.Close()
+}
+
+func (ps *GoRedisPubSub) Publish(ctx context.Context, topic, message string) error {
+	ps.mut.RLock()
+	defer ps.mut.RUnlock()
+	return ps.client.Publish(ctx, topic, message).Err()
+}
 
+func (ps *GoRedisPubSub) Subscribe(ctx context.Context, topic string) Subscription {
+	ps.mut.Lock()
+	defer ps.mut.Unlock()
+	sub := &GoRedisSubscription{
+		topic:  topic,
+		pubsub: ps.client.Subscribe(ctx, topic),
+		ch:     make(chan string, 100),
+		done:   make(chan struct{}),
+	}
+	ps.subs = append(ps.subs, sub)
 	go func() {
+		redisch := sub.pubsub.Channel()
 		for {
 			select {
-			case msg := <-t.redisSub.Channel():
-				t.mut.RLock()
-				subscribers := t.subscribers
-				t.mut.RUnlock()
-				for _, sub := range subscribers {
-					select {
-					case sub <- msg.Payload:
-					case <-t.done:
-						return
-					}
+			case <-sub.done:
+				close(sub.ch)
+				return
+			case msg := <-redisch:
+				if msg == nil {
+					continue
 				}
-			case <-t.done:
-				t.redisSub.Close()
-				t.mut.RLock()
-				subscribers := t.subscribers
-				for _, sub := range subscribers {
-					close(sub)
+				select {
+				case sub.ch <- msg.Payload:
+				default:
 				}
-				t.subscribers = nil
-				t.mut.RUnlock()
-				return
 			}
 		}
 	}()
-
-	ps.mut.Lock()
-	ps.topics[topic] = t
-	ps.mut.Unlock()
-	return t
-}
-
-// Close shuts down all topics and the redis connection
-func (ps *GoRedisPubSub) Close() {
-	ps.mut.Lock()
-	topics := maps.Values(ps.topics)
-	ps.mut.Unlock()
-
-	for _, t := range topics {
-		t.Close()
-	}
-	ps.client.Close()
-}
-
-// Publish sends a message to all subscribers of the topic
-func (t *GoRedisTopic) Publish(ctx context.Context, message string) error {
-	err := t.client.Publish(ctx, t.topic, message).Err()
-	if err != nil {
-		return err
-	}
-	return nil
+	return sub
 }
 
-// Subscribe returns a channel that will receive all messages published to the topic
-func (t *GoRedisTopic) Subscribe(ctx context.Context) <-chan string {
-	ch := make(chan string)
-	t.mut.Lock()
-	t.subscribers = append(t.subscribers, ch)
-	t.mut.Unlock()
-	return ch
+func (s *GoRedisSubscription) Channel() <-chan string {
+	return s.ch
 }
 
-// Close shuts down the topic and unsubscribes all subscribers
-func (t *GoRedisTopic) Close() {
-	t.once.Do(func() {
-		close(t.done)
-	})
+func (s *GoRedisSubscription) Close() {
+	s.pubsub.Close()
+	close(s.done)
```

Comment:
We probably need a `sync.Once` here for `close(s.done)` in case a client called `Close` and another goroutine called `GoRedisPubSub.Close()`